core/kotlinx-coroutines-debug/src/debug/internal/DebugProbesImpl.kt

```diff
@@ -0,0 +1,282 @@
+        } else {
+            val element = state.lastObservedStackTrace().firstOrNull()
+            val contState = state.state
+            builder.append("$str, continuation is $contState at line $element\n")
+        }
+
+        for (child in children) {
+            child.build(map, builder, indent + "\t")
+        }
+    }
+
+    @Synchronized
+    public fun dumpCoroutinesState(): List<CoroutineState> {
+        if (!isInstalled) {
+            error("Debug probes are not installed")
+        }
+
+        return capturedCoroutines.entries.asSequence()
+            .map { CoroutineState(it.key.delegate, it.value) }
+            .sortedBy { it.sequenceNumber }
+            .toList()
+    }
+
+    @Synchronized
+    public fun dumpCoroutines(out: PrintStream) {
+        if (!isInstalled) {
+            error("Debug probes are not installed")
+        }
+
+        // Avoid inference with other out/err invocations
+        val resultingString = buildString {
+            append("Coroutines dump ${dateFormat.format(System.currentTimeMillis())}")
+
+            capturedCoroutines
+                .asSequence()
+                .sortedBy { it.value.sequenceNumber }
+                .forEach { (key, value) ->
+                val state = if (value.state == State.RUNNING)
+                    "${value.state} (Last suspension stacktrace, not an actual stacktrace)"
+                else value.state.toString()
+
+                append("\n\nCoroutine $key, state: $state")
+                val observedStackTrace = value.lastObservedStackTrace()
+                if (observedStackTrace.isEmpty()) {
+                    append("\n\tat ${artificialFrame(ARTIFICIAL_FRAME_MESSAGE)}")
+                    printStackTrace(value.creationStackTrace)
+                } else {
+                    printStackTrace(value.lastObservedStackTrace())
+                }
+            }
+        }
+
+        // Move it out of synchronization?
+        out.println(resultingString)
+    }
+
+    private fun StringBuilder.printStackTrace(frames: List<StackTraceElement>) {
+        frames.forEach { frame ->
+            append("\n\tat $frame")
+        }
+    }
+
+    @Synchronized
+    internal fun probeCoroutineResumed(frame: Continuation<*>) = updateState(frame, State.RUNNING)
+
+    @Synchronized
+    internal fun probeCoroutineSuspended(frame: Continuation<*>) = updateState(frame, State.SUSPENDED)
+
+    private fun updateState(frame: Continuation<*>, state: State) {
+        if (!isInstalled) {
+            return
+        }
+
+        // Find ArtificialStackFrame of the coroutine
+        val owner = frame.owner()
+        val coroutineState = capturedCoroutines[owner]
+        if (coroutineState == null) {
+            warn(frame, state)
+            return
+        }
+
+        coroutineState.updateState(state, frame)
+    }
+
+    private fun Continuation<*>.owner(): ArtificialStackFrame<*>? {
+        var frame = this as? CoroutineStackFrame ?: return null
```

Comment:
Don't use `var`/`while` for things like that. Is is much more readable with tailrec:

```
    private fun Continuation<*>.owner(): ArtificialStackFrame<*>? =
        (this as? CoroutineStackFrame)?.owner()
        
    private tailrec fun CoroutineStackFrame.owner(): ArtificialStackFrame<*>? = 
        if (this is ArtificialStackFrame<*>) this else callerFrame?.owner()
```